paddle/fluid/framework/ir/shrink_assign_depvar_pass.cc

```diff
@@ -0,0 +1,130 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <stdio.h>
+#include <algorithm>
+#include <utility>
+
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/shrink_assign_depvar_pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+std::vector<ir::Node*> GetUpstreamOps(const ir::Node* op) {
+  std::vector<ir::Node*> upstreams;
+  for (ir::Node* var : op->inputs) {
+    for (ir::Node* upstream : var->inputs) {
+      PADDLE_ENFORCE(upstream->NodeType() == ir::Node::Type::kOperation);
+      upstreams.push_back(upstream);
+    }
+  }
+  return upstreams;
+}
+
+std::unique_ptr<ir::Graph> ShrinkAssignDepvarPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  std::vector<ir::Node*> to_remove;
+  //                       op      input var
+  std::vector<std::pair<ir::Node*, ir::Node*>> to_remove_unused_input;
+
+  for (ir::Node* n : graph->Nodes()) {
+    if (n->NodeType() == ir::Node::Type::kVariable) continue;
+    for (ir::Node* in_var : n->inputs) {
+      // FIXME(typhoonzero): shouldn't got op here?
+      if (in_var->NodeType() == ir::Node::Type::kOperation) continue;
+      for (ir::Node* upstream_op : in_var->inputs) {
+        if (upstream_op->Name() == "assign") {
+          VLOG(3) << "find path: " << upstream_op->Name() << "->"
+                  << in_var->Name() << "->" << n->Name();
+          // NOTE: This is a special case!!!
+          // connected by dep_var, cases like:
+          // condition_op -> learning_rate
+          //              \-> assign -> final_learning_rate ->
+          //                        \-> dep_var -> contition_op
+          // then, use dep_var to connect ops around assign
+          if (IsControlDepVar(*in_var)) {
+            to_remove.push_back(upstream_op);
+            to_remove_unused_input.push_back(
+                std::make_pair(in_var, upstream_op));
+            // add grand -> dep_var connect
+            auto before_assign_ops = GetUpstreamOps(upstream_op);
```

Comment:
It seems that line 63 is unnecessary, because the valid input of `assign` is one.